pkg/server/endpoints/registration/handler.go

```diff
@@ -545,6 +547,95 @@ func (h *Handler) ListAgents(ctx context.Context, listReq *registration.ListAgen
 	return &registration.ListAgentsResponse{Nodes: resp.Nodes}, nil
 }
 
+func (h *Handler) MintX509SVID(ctx context.Context, req *registration.MintX509SVIDRequest) (_ *registration.MintX509SVIDResponse, err error) {
+	counter := telemetry_registrationapi.StartMintX509SVIDCall(h.Metrics)
+	addCallerIDLabel(ctx, counter)
+	defer counter.Done(&err)
+
+	spiffeID, err := h.normalizeSPIFFEIDForMinting(req.SpiffeId)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(req.Csr) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "request missing CSR")
+	}
+
+	for _, dnsName := range req.DnsNames {
+		if err := validateDNS(dnsName); err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "%q is not a valid DNS name: %v", dnsName, err)
+		}
+	}
+
+	csr, err := x509.ParseCertificateRequest(req.Csr)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid CSR: %v", err)
+	}
+	if err := csr.CheckSignature(); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid CSR: signature verify failed")
+	}
+
+	svid, err := h.ServerCA.SignX509SVID(ctx, ca.X509SVIDParams{
+		SpiffeID:  spiffeID,
+		PublicKey: csr.PublicKey,
+		TTL:       time.Duration(req.Ttl) * time.Second,
+		DNSList:   req.DnsNames,
+	})
+
+	resp, err := h.getDataStore().FetchBundle(ctx, &datastore.FetchBundleRequest{
+		TrustDomainId: h.TrustDomain.String(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	if resp.Bundle == nil {
+		return nil, errors.New("bundle not found")
+	}
+
+	svidChain := make([][]byte, 0, len(svid))
+	for _, cert := range svid {
+		svidChain = append(svidChain, cert.Raw)
+	}
+
+	var rootCAs [][]byte
+	for _, rootCA := range resp.Bundle.RootCas {
+		rootCAs = append(rootCAs, rootCA.DerBytes)
+	}
+
+	return &registration.MintX509SVIDResponse{
+		SvidChain: svidChain,
+		RootCas:   rootCAs,
+	}, nil
+}
+
+func (h *Handler) MintJWTSVID(ctx context.Context, req *registration.MintJWTSVIDRequest) (_ *registration.MintJWTSVIDResponse, err error) {
+	counter := telemetry_registrationapi.StartMintJWTSVIDCall(h.Metrics)
+	addCallerIDLabel(ctx, counter)
+	defer counter.Done(&err)
+
+	spiffeID, err := h.normalizeSPIFFEIDForMinting(req.SpiffeId)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(req.Audience) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "request must specify at least one audience")
+	}
+
+	token, err := h.ServerCA.SignJWTSVID(ctx, ca.JWTSVIDParams{
+		SpiffeID: spiffeID,
+		TTL:      time.Duration(req.Ttl) * time.Second,
+		Audience: req.Audience,
+	})
+	if err != nil {
+		return nil, err
```

Comment:
grpc error code internal? any chance sensitive data can be returned here?

Reply:
internal is probably ok. i don't think any sensitive information is returned from here.